pallets/xcm-asset-manager/src/lib.rs

```diff
@@ -0,0 +1,448 @@
+	#[pallet::getter(fn asset_type_id)]
+	pub type AssetTypeId<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::ForeignAssetType, T::AssetId>;
+
+	/// Litentry: Purestake use AssetType as index. This might lead to security issue if
+	/// MultiLocation Hash changes. Here we use AssetId instead. Stores the units per second for
+	/// local execution for a AssetType. This is used to know how to charge for XCM execution in a
+	/// particular asset
+	/// Not all assets might contain units per second, hence the different storage
+	#[pallet::storage]
+	#[pallet::getter(fn asset_id_units_per_second)]
+	pub type AssetIdUnitsPerSecond<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, u128>;
+
+	/// Stores the tracker of foreign assets id that have been
+	/// created so far
+	#[pallet::storage]
+	#[pallet::getter(fn foreign_asset_tracker)]
+	pub type ForeignAssetTracker<T: Config> = StorageValue<_, T::AssetId, ValueQuery>;
+
+	/// Litentry: Purestake use AssetType as index. This might lead to security issue if
+	/// MultiLocation Hash changes. Here we use AssetId instead.
+	// Supported fee asset payments
+	#[pallet::storage]
+	#[pallet::getter(fn supported_fee_payment_assets)]
+	pub type SupportedFeePaymentAssets<T: Config> = StorageValue<_, Vec<T::AssetId>, ValueQuery>;
+
+	/// The storages for AssetMetadatas.
+	///
+	/// AssetMetadatas: map AssetId => Option<AssetMetadata>
+	#[pallet::storage]
+	#[pallet::getter(fn asset_metadatas)]
+	pub type AssetMetadatas<T: Config> =
+		StorageMap<_, Twox64Concat, T::AssetId, AssetMetadata<BalanceOf<T>>, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register new asset with the asset manager
+		#[pallet::weight(T::WeightInfo::register_foreign_asset())]
+		pub fn register_foreign_asset(
+			origin: OriginFor<T>,
+			asset_type: T::ForeignAssetType,
+			metadata: AssetMetadata<BalanceOf<T>>,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			let asset_id = ForeignAssetTracker::<T>::get();
+
+			ensure!(AssetTypeId::<T>::get(&asset_type).is_none(), Error::<T>::AssetAlreadyExists);
+
+			AssetMetadatas::<T>::insert(&asset_id, &metadata);
+			AssetIdType::<T>::insert(&asset_id, &asset_type);
+			AssetTypeId::<T>::insert(&asset_type, &asset_id);
+
+			Self::deposit_event(Event::<T>::ForeignAssetRegistered { asset_id, asset_type });
+			Self::deposit_event(Event::<T>::ForeignAssetMetadataUpdated { asset_id, metadata });
+
+			// Auto increment for Asset cunter
+			ForeignAssetTracker::<T>::put(asset_id + One::one());
+
+			Ok(())
+		}
+
+		/// Relocate asset id.
+		/// Can only be larger than current assignment.
+		#[pallet::weight(T::WeightInfo::relocate_foreign_asset_id())]
+		pub fn relocate_foreign_asset_id(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+		) -> DispatchResult {
+			T::ForeignAssetModifierOrigin::ensure_origin(origin)?;
+
+			if asset_id > ForeignAssetTracker::<T>::get() {
+				ForeignAssetTracker::<T>::put(asset_id);
+			}
```

Comment:
We need an event and/or error here, otherwise we won't know if this call actually changed the state as it always returns `Ok`

Reply:
Event ForeignAssetTrackerUpdated added if the change does happen.